File: CPPSolutions/CPPSolutions/M388_文件的最长绝对路径.cpp
using namespace std;
#include <cstdio>
#include <string>
#define MAX 10010

class Solution {
public:
    int dirs[MAX];
    int lengthLongestPath(string input) {
        memset(dirs, -1, sizeof(dirs));
        int n = input.length();
        int ans = 0;

        int i = 0;
        while (i < n) {
            int level = 0;
            while (i < n && input[i] == '\t') {
                level++;
                i++;
            }
            int j = i;
            bool isDir = 1;
            while (j < n && input[j] != '\n') {
                if (input[j++] == '.') {
                    isDir = 0;
                }
            }
            int cur = j - i;
            int prev = level - 1 >= 0 ? dirs[level-1] : -1;
            int path = prev + 1 + cur;
            if (isDir) {
                dirs[level] = path;
            }
            else if (path > ans) {
                ans = path;
            }
            i = j + 1;
        }
        return ans;
    }
};

File: LeetCode/src/Graph/DFS/M388_文件的最长绝对路径.java
package Graph.DFS;

import java.util.Arrays;

/**
 * @author Hyperspace
 * @date 2022/04/20
 */
public class M388_文件的最长绝对路径 {
    static final int MAX = 10000 + 10;
    static int[] dirs = new int[MAX];

    public int lengthLongestPath(String input) {
        Arrays.fill(dirs, -1);
        int n = input.length();
        int ans = 0;

        int i = 0;
        while (i < n) {
            int level = 0;
            while (i < n && input.charAt(i) == '\t') {
                level++;
                i++;
            }
            int j = i;
            boolean isDir = true;
            while (j < n && input.charAt(j) != '\n') {
                if (input.charAt(j++) == '.') {
                    isDir = false;
                }
            }
            int cur = j - i;
            int prev = level - 1 >= 0 ? dirs[level - 1] : -1;
            int path = prev + 1 + cur;
            if (isDir) {
                dirs[level] = path;
            } else if (path > ans) {
                ans = path;
            }
            i = j + 1;
        }

        return ans;
    }
}
